Src/Newtonsoft.Json/Utilities/AsyncUtils.cs

```diff
@@ -0,0 +1,113 @@
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+#endregion
+
+#if !(NET20 || NET35 || NET40 || PORTABLE40)
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Newtonsoft.Json.Utilities
+{
+    internal static class AsyncUtils
+    {
+        private static readonly Action Nop = () =>
+        {
+        };
+
+        // Pre-allocate to avoid wasted allocations.
+        public static readonly Task<bool> False = Task.FromResult(false);
+        public static readonly Task<bool> True = Task.FromResult(true);
+
+        internal static Task<bool> ToAsync(this bool value) => value ? True : False;
+
+        public static Task CancelIfRequesedAsync(this CancellationToken cancellationToken)
+        {
+            return cancellationToken.IsCancellationRequested ? CancelledAsync(cancellationToken) : null;
+        }
+
+        public static Task<T> CancelIfRequesedAsync<T>(this CancellationToken cancellationToken)
+        {
+            return cancellationToken.IsCancellationRequested ? CancelledAsync<T>(cancellationToken) : null;
+        }
+
+        public static Task CancelledAsync(this CancellationToken cancellationToken)
+        {
+            Debug.Assert(cancellationToken.IsCancellationRequested);
+            return new Task(Nop, cancellationToken);
+        }
+
+        public static Task<T> CancelledAsync<T>(this CancellationToken cancellationToken)
+        {
+            Debug.Assert(cancellationToken.IsCancellationRequested);
+            return new Task<T>(() => default(T), cancellationToken);
+        }
+
+        // Task.Delay(0) is optimised as a cached task within the framework, and indeed
+        // the same cached task that Task.CompletedTask returns as of 4.6, but we'll add
+        // our own property for previous frameworks.
+        internal static Task CompletedTask => Task.Delay(0);
```

Comment:
@ayende the caching within `Task.Delay()` already avoids allocation for the case of `0`, so on that count it's not necessary, but it will be a faster call as `Task.Delay(0)` hits a range check and a check on cancellation being needed that caching will skip.